### DISCUSSION TITLE
Support interoperability with Privy (and other wallets)

### DISCUSSION THREAD

**Discussion by akoita** [ProjectOpenSea/opensea-js#1357] [General]
Hello!

After switching to version 7 of opensea-js, interoperability with providers of Privy (which we use) no longer works, and this is blocking us.

Privy only supports Ethers.js version 5.7. And the other providers it supplies (Web3js and EIP1193 providers) don't work with opensea-js version 7 either. 

Although, the EIP1193 provider seemed to be working fine, until I encountered an error "TypeError: signer.signTypedData is not a function at Seaport.signOrder".

Is it possible for you to add support for other providers to expand the possibility of interacting with your client?

**Integrating Privy with third-party libraries:**

- https://docs.privy.io/guide/frontend/embedded/usage/3p-libraries
- https://docs.privy.io/guide/frontend/wallets/external

Thank you

**Comment by ryanio** [comment]
You'll need to use ethers v6 with opensea-js v7. It should be okay to use both ethers v5 and v6 for the meantime if you need to use v5 with other packages.

**Comment by akoita** [comment]
@ryanio Thanks for the help, but it doesn't work. More info on our use case:

- **@privy-io/react-auth**:  1.52.1
- **opensea-js**: 7.0.7

Privy is the embedded wallet of our app. So to instantiate opensea-js SDK, we proceed as follows: 

```
const provider = await embeddedWallet.getEthersProvider(); // ethers provider object from Privy
const openseaSDK = new OpenSeaSDK(
                provider,
                {
                    chain: chainName,
                    apiKey: API_KEY,
                },
                openseaLogger
            );
```

This code worked well (and still works) with opensea-js-6.x. But when we switched to opensea-js-7, it no longer works with this Privy provider. Privy only supports Ethers.js v5, so we have no control over the provider version we pass to opensea-js: https://docs.privy.io/guide/frontend/embedded/usage/3p-libraries

The error I had at the beginning was this one: 
```
Specified accountAddress is not available through wallet or provider: <privy wallet address>. . Accounts available:
    at a7r._requireAccountIsAvailable (sdk.js:609:15)
    at async a7r.createListing (sdk.js:228:9)
    at async z7e.listToken (opensea.js:40:33)
    at async onSubmit (NFTListModalButton.jsx:51:29)

```

I then modified the SDK instantiation as follows: 

```

            const provider = await embeddedWallet.getEthersProvider(); // ethers provider object  from Privy
            const walletProvider = provider.provider.walletProvider;

            const openseaSDK = new OpenSeaSDK(
                walletProvider,
                {
                    chain: chainName,
                    apiKey: API_KEY,
                },
                openseaLogger
            );

```

With this instantiation, execution goes further, but ends up crashing with the error: 
```
Error: processing response error (body="{\"jsonrpc\":\"2.0\",\"error\":{\"code\":-32000,\"message\":\"unknown account\"},\"id\":54}\n", error={"code":-32000}, requestBody="{\"method\":\"eth_sendTransaction\",\"params\":[{\"gas\":\"0xb50a\",\"from\":\"<privy wallet address>\",\"to\":\"0x2b6909416fbb2c27ef51b14dc7897d410e33eb86\",\"data\":\"0xa22cb4650000000000000000000000001e0049783f008a0085193e00003d00cd54003c710000000000000000000000000000000000000000000000000000000000000001\"}],\"id\":54,\"jsonrpc\":\"2.0\"}", requestMethod="POST", url="https://sepolia.base.org", code=SERVER_ERROR, version=web/5.7.1)
    at _Logger.makeError (chunk-Z62LSFF7.js?v=372c4adb:987:19)
    at _Logger.throwError (chunk-Z62LSFF7.js?v=372c4adb:996:16)
    at chunk-Z62LSFF7.js?v=372c4adb:6727:22
    at Generator.next (<anonymous>)
    at fulfilled (chunk-Z62LSFF7.js?v=372c4adb:6486:24)
```



Privy provides an lll provider, but using it I get the same error as above:

```
const provider = await embeddedWallet.getEthereumProvider()  // EIP1193 provider from Privy
const openseaSDK = new OpenSeaSDK(
                provider,
                {
                    chain: chainName,
                    apiKey: API_KEY,
                },
                openseaLogger
            );

```


Having reached an impasse, we decided to switch back to opensea-js-6.x, but this version unfortunately does not support BaseSepolia ☹️

I'm currently stuck.  If you have an idea for a solution, it would help a lot. 
Thank you

**Reply by ryanio** [reply]
Thanks for the details. 

You should be able to fork opensea-js v6 and add base sepolia to the chain enum, but not all the functions will work if you use any that call the v1 api 

Question is also if you can wrap the provider returned by privy in a ethers v6 provider to use with opensea-js v7

**Reply by akoita** [reply]
Hi @ryanio,
We've added support for BaseSepolia in version 6.x. Is it possible and interesting to submit a PR of this change?
```
From 1a537d83dfd7ba8d57e8858369febf4ebed2f0e4 Mon Sep 17 00:00:00 2001
From:  <>
Date: Mon, 15 Jan 2024 11:18:34 +0100
Subject: [PATCH] Add BaseSepolia support

---
 src/types.ts       | 6 +++++-
 src/utils/utils.ts | 4 ++--
 2 files changed, 7 insertions(+), 3 deletions(-)

diff --git a/src/types.ts b/src/types.ts
index e5f8bba9e..7e1267e82 100644
--- a/src/types.ts
+++ b/src/types.ts
@@ -45,6 +45,10 @@ export enum EventType {
    * Emitted when the {@link OpenSeaSDK.approveOrder} method is called.
    */
   ApproveOrder = "ApproveOrder",
+  /**
+   * Emitted when the {@link OpenSeaSDK.transfer} method is called.
+   */
+  Transfer = "Transfer",
 }
 
 /**
@@ -131,7 +135,7 @@ export enum Chain {
   /** Klaytn Baobab */
   Baobab = "baobab",
   /** Base L2 Testnet */
-  BaseGoerli = "base_goerli",
+  BaseSepolia = "base_sepolia",
   /** Binance Smart Chain Testnet */
   BNBTestnet = "bsctestnet",
   /** Arbitrum Testnet */
diff --git a/src/utils/utils.ts b/src/utils/utils.ts
index de67386a7..46a7e5207 100644
--- a/src/utils/utils.ts
+++ b/src/utils/utils.ts
@@ -347,7 +347,7 @@ export const getWETHAddress = (chain: Chain) => {
       return "0xae13d989dac2f0debff460ac112a837c89baa7cd";
     // OP Chains have weth at the same address
     case Chain.Base:
-    case Chain.BaseGoerli:
+    case Chain.BaseSepolia:
     case Chain.Optimism:
     case Chain.OptimismGoerli:
     case Chain.Zora:
@@ -402,7 +402,7 @@ export const isTestChain = (chain: Chain): boolean => {
     case Chain.Sepolia:
     case Chain.Mumbai:
     case Chain.Baobab:
-    case Chain.BaseGoerli:
+    case Chain.BaseSepolia:
     case Chain.BNBTestnet:
     case Chain.ArbitrumSepolia:
     case Chain.Fuji:
```

**Reply by ryanio** [reply]
Great to hear, thanks! I don't think that's necessary we don't really support prior versions but thanks for sharing the branch if others need it they can find it here.

**Reply by adrienthiery** [reply]
Hi @ryanio , so following last week's killing of the v1 API, we did upgrade to opensea-js v7 and managed to fix the incompatibility issue with privy's exported provider by transitioning through the EIP1193 provider:

```
const ethereum = await embeddedWallet.getEthereumProvider();
const ethers6Provider = new BrowserProvider(ethereum);

const openseaSDK = new OpenSeaSDK(
    ethers6Provider,
    {
        chain: chainName,
        apiKey: API_KEY,
    }
);
```

For anyone that might be interested!
